File: code/scala/src/main/scala/group/pals/desktop/app/apksigner/utils/Signer.scala
/*
 *    Copyright (C) 2012 Hai Bison
 *
 *    See the file LICENSE at the root directory of this project for copying
 *    permission.
 */

package group.pals.desktop.app.apksigner.utils

import group.pals.desktop.app.apksigner.i18n.Messages
import group.pals.desktop.app.apksigner.i18n.R

import java.io.File
import java.io.IOException
import java.io.InputStream
import java.util.regex.Matcher

/**
 * Helper class to sign APK files.
 *
 * @author Hai Bison
 */
object Signer {

    /**
     * Used to append to newly signed target's file name.
     */
    private lazy final val SIGNED = "SIGNED"

    /**
     * Used to append to newly signed target's file name.
     */
    private lazy final val SIGNED_UNALIGNED = "SIGNED_UNALIGNED"

    /**
     * Signs an APK file.
     *
     * @param jdkPath
     *            the path to JDK, can be {@code null} on Unix system.
     * @param targetFile
     *            the target file, can be an APK, JAR or ZIP.
     * @param keyFile
     *            the keystore file.
     * @param storepass
     *            the keystore's password.
     * @param alias
     *            the keystore alias.
     * @param keypass
     *            the keystore's alias password.
     * @return
     * @throws IOException
     *             if any occurred.
     * @throws InterruptedException
     *             if any occurred.
     */
    def sign(jdkPath: File, targetFile: File, keyFile: File,
            storepass: Array[Char], alias: String,
            keypass: Array[{Char]): String = {

        /*
         * JDK for Linux does not need to specify full path
         */
        var jarsigner =
            if (jdkPath != null && jdkPath.isDirectory())
                jdkPath.getAbsolutePath() + "/jarsigner.exe"
            else "jarsigner"

        /*
         * jarsigner -keystore KEY_FILE -sigalg MD5withRSA -digestalg SHA1
         * -storepass STORE_PASS -keypass KEY_PASS APK_FILE ALIAS_NAME
         */
        var pb = new ProcessBuilder(Array(
                jarsigner,
                "-keystore", keyFile.getAbsolutePath(), "-sigalg",
                "MD5withRSA", "-digestalg", "SHA1", "-storepass",
                new String(storepass), "-keypass", new String(keypass),
                targetFile.getAbsolutePath(), alias))
        var p = pb.start()

        var console = new StringBuilder()
        var stream = p.getInputStream()
        try {
            var read = 0
            var buf = new Array[Byte](Files.FILE_BUFFER)
            while ({ read = stream.read(buf); read } > 0)
                console.append(new String(buf, 0, read))
        } finally {
            if (stream != null)
                stream.close()
        }

        /*
         * TODO: parse output for errors, warnings...
         */

        p.waitFor()

        /*
         * Renames newly signed file...
         */

        val result = console.toString().trim()
        if (result.isEmpty()) {
            val oldName = targetFile.getName()
            var newName: String = null
            if (oldName.matches("(?si).*?unsigned.+")) {
                if (oldName.matches(Texts.REGEX_APK_FILES))
                    newName = oldName.replaceFirst("(?si)unsigned",
                            Matcher.quoteReplacement(SIGNED_UNALIGNED))
                else
                    newName = oldName.replaceFirst("(?si)unsigned",
                            Matcher.quoteReplacement(SIGNED))
            } else if (oldName.matches(Texts.REGEX_APK_FILES))
                newName = Files.appendFilename(oldName, '_' + SIGNED_UNALIGNED)
            else if (oldName.matches(Texts.REGEX_JAR_FILES)
                    || oldName.matches(Texts.REGEX_ZIP_FILES))
                newName = Files.appendFilename(oldName, '_' + SIGNED)
            else
                newName = String.format("%s_%s", oldName, SIGNED)

            if (targetFile.renameTo(new File(targetFile.getParent()
                    + File.separator + newName)))
                return null

            Messages.getString(
                R.string.pmsg_file_is_signed_but_cannot_be_renamed_to_new_one,
                newName)
        }// results from console is empty
        else {
            result
        }// results from console is NOT empty
    }// sign()

}

File: code/jython/apksigner/utils/signer.py
#  Copyright (C) 2012 Hai Bison
#
#  See the file LICENSE at the root directory of this project for copying
#  permission.

'''
Tool to sign APK files.
'''

import os, re, subprocess

import files, texts

from apksigner.i18n import messages, string

'''
Used to append to newly signed target's file name.
'''
SIGNED = 'SIGNED'

'''
Used to append to newly signed target's file name.
'''
SIGNED_UNALIGNED = 'SIGNED_UNALIGNED'

def sign(jdk_path, target_file, key_file, storepass, alias, keypass):
    ''' Signs an APK file.

        Parameters:

        :jdk_path:
            the path to JDK's `bin` directory, can be ``None`` on Unix system.
        :target_file:
            the target file, can be an APK, JAR or ZIP.
        :key_file:
            the keystore file.
        :storepass:
            the keystore's password.
        :alias:
            the keystore alias.
        :keypass:
            the keystore's alias password.

        Returns:
            ``None`` if everything is OK, or an error message if it occurred.
    '''

    # JDK for Linux does not need to specify full path
    jarsigner = 'jarsigner'
    if jdk_path and os.path.isdir(jdk_path):
        jarsigner = os.path.join(jdk_path, jarsigner + '.exe')

    # jarsigner -keystore KEY_FILE -sigalg MD5withRSA -digestalg SHA1
    # -storepass STORE_PASS -keypass KEY_PASS APK_FILE ALIAS_NAME
    cmd = [ jarsigner, '-keystore', key_file, '-sigalg', 'MD5withRSA',
            '-digestalg', 'SHA1', '-storepass', storepass, '-keypass', keypass,
            target_file, alias ]
    try:
        output = subprocess.check_output(cmd)
    except Exception as e:
        return e.output

    if not output: return output

    ### RENAMES NEWLY SIGNED FILE...

    OLD_NAME = os.path.basename(target_file)
    if re.match(r'(?si).*?unsigned.+', OLD_NAME):
        if re.match(texts.REGEX_APK_FILES, OLD_NAME):
            new_name = re.sub(r'(?si)unsigned', SIGNED_UNALIGNED, OLD_NAME, 1)
        else:
            new_name = re.sub(r'(?si)unsigned', SIGNED, OLD_NAME, 1)
    elif re.match(texts.REGEX_APK_FILES, OLD_NAME):
        new_name = files.append_filename(OLD_NAME, '_' + SIGNED_UNALIGNED)
    elif re.match(texts.REGEX_JAR_FILES, OLD_NAME) or \
        re.match(texts.REGEX_ZIP_FILES, OLD_NAME):
        new_name = files.append_filename(OLD_NAME, '_' + SIGNED)
    else:
        new_name = '{}_{}'.format(OLD_NAME, SIGNED)

    new_name = os.path.join(os.path.dirname(target_file), new_name)
    try:
        os.rename(target_file, new_name)
    except:
        return messages.get_string(
            string.pmsg_file_is_signed_but_cannot_be_renamed_to_new_one, new_name)
    #.sign()